File: src/main/java/org/jukeboxmc/block/BlockTwistingVines.java
package org.jukeboxmc.block;

import org.jukeboxmc.item.ItemTwistingVines;

public class BlockTwistingVines extends Block {

    public BlockTwistingVines() {
        super("minecraft:twisting_vines");
    }

    @Override
    public ItemTwistingVines toItem() {
        return new ItemTwistingVines();
    }

    @Override
    public BlockType getBlockType() {
        return BlockType.TWISTING_VINES;
    }

    @Override
    public boolean isSolid() {
        return false;
    }

    @Override
    public boolean isTransparent() {
        return true;
    }

    public void setTwistingVinesAge( int value ) { //0-25
        this.setState( "twisting_vines_age", value );
    }

    public int getTwistingVinesAge() {
        return this.stateExists( "twisting_vines_age" ) ? this.getIntState( "twisting_vines_age" ) : 0;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockTwistingVines.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.TwistingVines
import org.jukeboxmc.api.item.Item
import org.jukeboxmc.api.item.ToolType
import org.jukeboxmc.api.item.enchantment.EnchantmentType
import org.jukeboxmc.server.block.JukeboxBlock

class BlockTwistingVines(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    TwistingVines {

    override fun canPassThrough(): Boolean {
        return true
    }

    override fun getTwistingVinesAge(): Int {
        return this.getIntState("twisting_vines_age")
    }

    override fun setTwistingVinesAge(value: Int): TwistingVines {
        return this.setState("twisting_vines_age", value)
    }

    override fun getDrops(item: Item): MutableList<Item> {
        if (item.getToolType() == ToolType.SHEARS) {
            return mutableListOf(this.toItem())
        } else {
            var level = 0
            if (item.hasEnchantment(EnchantmentType.FORTUNE)) {
                level = item.getEnchantment(EnchantmentType.FORTUNE)!!.getLevel()
            }
            val probability = when (level) {
                0 -> {
                    33.0f
                }

                1 -> {
                    55.0f
                }

                2 -> {
                    77.0f
                }

                3 -> {
                    100.0f
                }

                else -> 0.0f
            }
            if (this.getRandom().nextFloat(100.0f) <= probability) {
                return mutableListOf(this.toItem())
            }
            return mutableListOf()
        }
    }
}